processing/src/main/java/io/druid/query/groupby/epinephelinae/SpillingGrouper.java

```diff
@@ -0,0 +1,198 @@
+{
+  private static final Logger log = new Logger(SpillingGrouper.class);
+
+  private final BufferGrouper<KeyType> grouper;
+  private final KeySerde<KeyType> keySerde;
+  private final LimitedTemporaryStorage temporaryStorage;
+  private final ObjectMapper spillMapper;
+  private final AggregatorFactory[] aggregatorFactories;
+
+  private final List<File> files = Lists.newArrayList();
+  private final List<Closeable> closeables = Lists.newArrayList();
+
+  public SpillingGrouper(
+      final ByteBuffer buffer,
+      final KeySerdeFactory<KeyType> keySerdeFactory,
+      final ColumnSelectorFactory columnSelectorFactory,
+      final AggregatorFactory[] aggregatorFactories,
+      final LimitedTemporaryStorage temporaryStorage,
+      final ObjectMapper spillMapper,
+      final int maxBufferGrouperSize
+  )
+  {
+    this.keySerde = keySerdeFactory.factorize();
+    this.grouper = new BufferGrouper<>(
+        buffer,
+        keySerde,
+        columnSelectorFactory,
+        aggregatorFactories,
+        maxBufferGrouperSize
+    );
+    this.aggregatorFactories = aggregatorFactories;
+    this.temporaryStorage = temporaryStorage;
+    this.spillMapper = spillMapper;
+  }
+
+  @Override
+  public boolean aggregate(KeyType key, int keyHash)
+  {
+    if (grouper.aggregate(key, keyHash)) {
+      return true;
+    } else {
+      // TODO(gianm): This blocks up all other processing threads, maybe split the buffer so we can spill in background
+      spill();
+      return grouper.aggregate(key, keyHash);
+    }
+  }
+
+  @Override
+  public boolean aggregate(KeyType key)
+  {
+    return aggregate(key, Groupers.hash(key));
+  }
+
+  @Override
+  public void reset()
+  {
+    grouper.reset();
+    deleteFiles();
+  }
+
+  @Override
+  public void close()
+  {
+    grouper.close();
+    deleteFiles();
+  }
+
+  @Override
+  public Iterator<Entry<KeyType>> iterator(final boolean sorted)
+  {
+    final List<Iterator<Entry<KeyType>>> iterators = new ArrayList<>(1 + files.size());
+
+    iterators.add(grouper.iterator(sorted));
+
+    for (final File file : files) {
+      final MappingIterator<Entry<KeyType>> fileIterator = read(file, keySerde.keyClazz());
+      iterators.add(
+          Iterators.transform(
+              fileIterator,
+              new Function<Entry<KeyType>, Entry<KeyType>>()
+              {
+                @Override
+                public Entry<KeyType> apply(Entry<KeyType> entry)
+                {
+                  final Object[] deserializedValues = new Object[entry.getValues().length];
+                  for (int i = 0; i < deserializedValues.length; i++) {
+                    deserializedValues[i] = aggregatorFactories[i].deserialize(entry.getValues()[i]);
+                    if (deserializedValues[i] instanceof Integer) {
+                      // TODO(gianm): This is a hack to satisfy the groupBy tests, maybe we can do better.
```

Comment:
the smileMapper deserializes small numbers as `Integer` and large numbers as `Long` but the groupBy tests expect all numbers to be `Long`. So this casts them.